File: Android/app/src/main/java/ru/inwords/inwords/domain/interactor/translation/TranslationSyncInteractorImpl.java
package ru.inwords.inwords.domain.interactor.translation;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.schedulers.Schedulers;
import ru.inwords.inwords.data.sync.TranslationSyncController;

public class TranslationSyncInteractorImpl implements TranslationSyncInteractor {
    private final TranslationSyncController syncController;

    @Inject
    TranslationSyncInteractorImpl(TranslationSyncController syncController) {
        this.syncController = syncController;
    }

    @Override
    public Completable presyncOnStart() {
        return syncController.presyncOnStart()
                .subscribeOn(Schedulers.io())
                .doOnError(Throwable::printStackTrace)
                .ignoreElement();
    }

    @Override
    public Completable trySyncAllReposWithCache() {
        return syncController.trySyncAllReposWithCache()
                .subscribeOn(Schedulers.io());
    }

    @Override
    public void notifyDataChanged() {
        syncController.notifyDataChanged();
    }
}

File: Android/app/src/main/java/ru/inwords/inwords/domain/interactor/translation/TranslationSyncInteractorImpl.kt
package ru.inwords.inwords.domain.interactor.translation

import android.util.Log
import io.reactivex.Completable
import ru.inwords.inwords.data.sync.TranslationSyncController
import javax.inject.Inject

class TranslationSyncInteractorImpl @Inject
internal constructor(private val syncController: TranslationSyncController) : TranslationSyncInteractor {

    override fun presyncOnStart(): Completable {
        return syncController.presyncOnStart()
                .doOnError { Log.e(javaClass.simpleName, it.message.orEmpty()) }
                .ignoreElement()
    }

    override fun trySyncAllReposWithCache(): Completable {
        return syncController.trySyncRemoteReposWithLocal()
    }

    override fun notifyDataChanged() = syncController.notifyDataChanged()
}
